### PR TITLE
config-linux: clarify the handling of ClosID RDT parameter

### PR DESCRIPTION
**Pull request by marquiz**
- An attempt to make the spec easier to interpret by grouping all ClosID related contraints in one place.
- fix indentation on IntelRdt, and split out the rules regarding interdependency of parameters into a separate list.

### CODE REVIEW

File: config-linux.md [hunk 1]
```diff
@@ -535,19 +535,21 @@ The following parameters can be specified for the container:
...
+* If neither `l3CacheSchema` nor `memBwSchema` is set, runtimes MUST NOT write to `schemata` files in any `resctrl` pseudo-filesystems.
 
-    * If `closID` is set, and neither of `l3CacheSchema` and `memBwSchema` are set, runtime MUST check if corresponding pre-configured directory `closID` is present in mounted `resctrl`. If such pre-configured directory `closID` exists, runtime MUST assign container to this `closID` and [generate an error](runtime.md#errors) if directory does not exist.
+* If `closID` is set, `l3CacheSchema` and/or `memBwSchema` is set, runtimes MUST compare `l3CacheSchema` and/or `memBwSchema` value with `schemata` file, and [generate an error](runtime.md#errors) if doesn't match.
+
+* If `closID` is set, and neither of `l3CacheSchema` and `memBwSchema` are set, runtime MUST check if corresponding pre-configured directory `closID` is present in mounted `resctrl`. If such pre-configured directory `closID` exists, runtime MUST assign container to this `closID` and [generate an error](runtime.md#errors) if directory does not exist.
```
Comment (kolyshkin) [inline config-linux.md:555]: > If such pre-configured directory `closID` exists, runtime MUST assign container to this `closID` and [generate an error](runtime.md#errors) if directory does not exist.

I read this as (in pseudocode):
```
if directory_exists() {
    assign(container, ClosID)
    if !directory_exists()
        error()
}
```

So maybe reword this to improve clarity.

Comment (kolyshkin) [reply config-linux.md:555]: Ah, you do this in the second commit. Nice!

**Review by kolyshkin — APPROVED** [review]
LGTM (non-binding)

**Comment by marquiz** [comment]
Thanks for the review @kolyshkin!

Is there somebody I should assign this to or smth else to do except for sit and wait? 😄

**Comment by marquiz** [comment]
ping

**Review by tianon — APPROVED** [review]
LGTM

**Comment by xiaochenshen** [comment]
LGTM.

Making use of pre-configured ClosIDs is a good improvement, it helps prevent running out of hardware CLOSIDs in container user scenarios. The number of CLOSIDs have a hardware limit which is exposed by kernel resctrl filesystem, and is also exposed by runc events command (GetStats):
cat /sys/fs/resctrl/info/L3/num_closids
cat /sys/fs/resctrl/info/MB/num_closids

**Comment by marquiz** [comment]
ping @caniszczyk @crosbymichael @cyphar @dqminh @giuseppe @hqhq @mrunalp @vbatts @vishh

**Review by giuseppe — APPROVED** [review]
LGTM

**Comment by marquiz** [comment]
ping @crosbymichael could this be merged?

**Comment by marquiz** [comment]
Is there a reason to keep this pending?

ping @crosbymichael @mrunalp @vbatts @dqminh @tianon @hqhq @cyphar @giuseppe

**Review by vbatts — APPROVED** [review]

**Comment by marquiz** [comment]
Thanks @vbatts 🚀 😄

File: config-linux.md [hunk 2]
```diff
@@ -530,24 +530,29 @@ If `intelRdt` is not set, the runtime MUST NOT manipulate any `resctrl` pseudo-f
...
+    The value MUST start with `MB:` and MUST NOT contain newlines.
 
-    * If both `l3CacheSchema` and `memBwSchema` are set, runtimes MUST write the combined value to the `schemata` file in that sub-directory discussed in `closID`.
+The following rules on parameters MUST be applied:
 
-    * If `l3CacheSchema` contains a line beginning with `MB:`, the value written to `schemata` file MUST be the non-`MB:` line(s) from `l3CacheSchema` and the line from `memBWSchema`.
+* If both `l3CacheSchema` and `memBwSchema` are set, runtimes MUST write the combined value to the `schemata` file in that sub-directory discussed in `closID`.
 
-    * If either `l3CacheSchema` or `memBwSchema` is set, runtimes MUST write the value to the `schemata` file in the that sub-directory discussed in `closID`.
+* If `l3CacheSchema` contains a line beginning with `MB:`, the value written to `schemata` file MUST be the non-`MB:` line(s) from `l3CacheSchema` and the line from `memBWSchema`.
```
Comment (giuseppe) [inline config-linux.md:543]: @marquiz I am looking at the specs to implement this in crun but I don't understand this rule.

Can you give an example?  Should it be dropped (at least I don't see any special handling in runc either for this)?

Comment (marquiz) [reply config-linux.md:543]: @giuseppe the idea is that if you have 
```
l3CacheSchema: "L3:foobar\nMB:abcd"
memBWSchema: "MB:xyzy"
```

Then you'd actually write `L3:foobar\nMB:xyzy` to the schemata file.

I don't think this has been implemented anywhere. I didn't write this particular formulation, and TBH, I wouldn't probably bother implementing it. I think the fields and there handling is just ambiguously specified, e.g. somehow it's implied that `l3CacheSchema` could contain multiple lines but `memBWSchema` only one line. What if l3CacheSchema contains `L3:...` but  memBWSchema: contains `L3:...` (and other similar questions)...

I try to address the problems in #1230

Comment (giuseppe) [reply config-linux.md:543]: thanks, I understand now.  Still feels weird to specify that  `MB:` is filtered out instead of just not expecting such a case :-) but I guess we can live with it at this point
